File: plugins/audio_player/android/module/src/main/java/net/chentong/leaf/android/plugin/audioplayer/AudioPlayerPlugin.java
package net.chentong.leaf.android.plugin.audioplayer;

import android.content.Context;
import android.media.AudioAttributes;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;

import net.chentong.leaf.android.plugin.LeafPlugin;

import org.json.JSONObject;

import java.util.ArrayDeque;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;

public class AudioPlayerPlugin implements LeafPlugin {
    private final Context context;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Map<String, PlayerRecord> players = new ConcurrentHashMap<>();

    public AudioPlayerPlugin(Context context) {
        this.context = context == null ? null : context.getApplicationContext();
    }

    @Override
    public String pluginName() {
        return "AudioPlayerPlugin";
    }

    @Override
    public boolean canHandle(String method) {
        return method != null && method.startsWith("audio_player.");
    }

    @Override
    public void onMethodCall(LeafMethodCall call, Result result) {
        String method = call.getMethod();
        try {
            if ("audio_player.set_source".equals(method)) {
                runOnMain(call, result, () -> handleSetSource(call, result));
                return;
            }
            if ("audio_player.play".equals(method)) {
                runOnMain(call, result, () -> handlePlay(call, result));
                return;
            }
            if ("audio_player.pause".equals(method)) {
                runOnMain(call, result, () -> handlePause(call, result));
                return;
            }
            if ("audio_player.stop".equals(method)) {
                runOnMain(call, result, () -> handleStop(call, result));
                return;
            }
            if ("audio_player.seek".equals(method)) {
                runOnMain(call, result, () -> handleSeek(call, result));
                return;
            }
            if ("audio_player.set_looping".equals(method)) {
                runOnMain(call, result, () -> handleSetLooping(call, result));
                return;
            }
            if ("audio_player.set_volume".equals(method)) {
                runOnMain(call, result, () -> handleSetVolume(call, result));
                return;
            }
            if ("audio_player.get_duration".equals(method)) {
                runOnMain(call, result, () -> handleGetDuration(call, result));
                return;
            }
            if ("audio_player.get_position".equals(method)) {
                runOnMain(call, result, () -> handleGetPosition(call, result));
                return;
            }
            if ("audio_player.is_playing".equals(method)) {
                runOnMain(call, result, () -> handleIsPlaying(call, result));
                return;
            }
            if ("audio_player.listen".equals(method)) {
                runOnMain(call, result, () -> handleListen(call, result));
                return;
            }
            if ("audio_player.dispose".equals(method)) {
                runOnMain(call, result, () -> handleDispose(call, result));
                return;
            }
            result.error(call.getRequestId(), -404, "method_not_implemented", false);
        } catch (Throwable throwable) {
            result.error(call.getRequestId(), -500, errorText(throwable), false);
        }
    }

    public void release() {
        mainHandler.post(() -> {
            for (PlayerRecord record : players.values()) {
                releaseRecord(record);
            }
            players.clear();
        });
    }

    private void runOnMain(LeafMethodCall call, Result result, ThrowingRunnable runnable) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            safeRun(call, result, runnable);
            return;
        }
        mainHandler.post(() -> safeRun(call, result, runnable));
    }

    private void safeRun(LeafMethodCall call, Result result, ThrowingRunnable runnable) {
        try {
            runnable.run();
        } catch (Throwable throwable) {
            result.error(call.getRequestId(), -500, errorText(throwable), false);
        }
    }

    private void handleSetSource(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        String playerId = requirePlayerId(args);
        String source = args.optString("source", "");
        if (source.isEmpty()) {
            result.error(call.getRequestId(), -1, "source_empty", false);
            return;
        }

        PlayerRecord record = getOrCreateRecord(playerId);
        preparePlayer(record, source);
        result.success(call.getRequestId(), buildPlayerJson(playerId).toString());
    }

    private void handlePlay(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        if (record.mediaPlayer == null) {
            result.error(call.getRequestId(), -2, "source_not_set", false);
            return;
        }
        record.mediaPlayer.start();
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handlePause(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        if (record.mediaPlayer != null && record.mediaPlayer.isPlaying()) {
            record.mediaPlayer.pause();
        }
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handleStop(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        if (record.mediaPlayer != null) {
            if (record.mediaPlayer.isPlaying()) {
                record.mediaPlayer.pause();
            }
            record.mediaPlayer.seekTo(0);
        }
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handleSeek(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        PlayerRecord record = requireRecord(args, call, result);
        if (record == null) return;
        double position = Math.max(0.0, args.optDouble("position", 0.0));
        if (record.mediaPlayer != null) {
            record.mediaPlayer.seekTo((int) Math.round(position * 1000.0));
        }
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handleSetLooping(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        PlayerRecord record = requireRecord(args, call, result);
        if (record == null) return;
        record.looping = args.optBoolean("looping", false);
        if (record.mediaPlayer != null) {
            record.mediaPlayer.setLooping(record.looping);
        }
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handleSetVolume(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        PlayerRecord record = requireRecord(args, call, result);
        if (record == null) return;
        double volume = args.optDouble("volume", 1.0);
        if (volume < 0.0) volume = 0.0;
        if (volume > 1.0) volume = 1.0;
        record.volume = (float) volume;
        if (record.mediaPlayer != null) {
            record.mediaPlayer.setVolume(record.volume, record.volume);
        }
        result.success(call.getRequestId(), buildPlayerJson(record.playerId).toString());
    }

    private void handleGetDuration(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        JSONObject out = buildPlayerJson(record.playerId);
        out.put("duration", durationSeconds(record));
        result.success(call.getRequestId(), out.toString());
    }

    private void handleGetPosition(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        JSONObject out = buildPlayerJson(record.playerId);
        out.put("position", positionSeconds(record));
        result.success(call.getRequestId(), out.toString());
    }

    private void handleIsPlaying(LeafMethodCall call, Result result) throws Exception {
        PlayerRecord record = requireRecord(call, result);
        if (record == null) return;
        JSONObject out = buildPlayerJson(record.playerId);
        out.put("playing", record.mediaPlayer != null && record.mediaPlayer.isPlaying());
        result.success(call.getRequestId(), out.toString());
    }

    private void handleListen(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        String playerId = requirePlayerId(args);
        PlayerRecord record = getOrCreateRecord(playerId);

        if (record.pendingListener != null) {
            result.error(call.getRequestId(), -11, "listen_in_progress", false);
            return;
        }

        if (!record.events.isEmpty()) {
            String event = record.events.poll();
            result.success(call.getRequestId(), event);
            return;
        }

        record.pendingListener = new PendingListener(call.getRequestId(), result);
    }

    private void handleDispose(LeafMethodCall call, Result result) throws Exception {
        JSONObject args = parseArgs(call.getArgs());
        String playerId = requirePlayerId(args);
        PlayerRecord record = players.remove(playerId);
        if (record != null) {
            releaseRecord(record);
        }
        result.success(call.getRequestId(), buildPlayerJson(playerId).toString());
    }

    private PlayerRecord getOrCreateRecord(String playerId) {
        PlayerRecord existing = players.get(playerId);
        if (existing != null) {
            return existing;
        }
        PlayerRecord record = new PlayerRecord();
        record.playerId = playerId;
        PlayerRecord raced = players.putIfAbsent(playerId, record);
        return raced == null ? record : raced;
    }

    private void preparePlayer(PlayerRecord record, String source) throws Exception {
        releaseMediaPlayerOnly(record);

        MediaPlayer player = new MediaPlayer();
        player.setAudioAttributes(new AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
            .build());
        player.setLooping(record.looping);
        player.setVolume(record.volume, record.volume);
        player.setOnCompletionListener(mp -> {
            if (record.looping) {
                return;
            }
            enqueueEvent(record, "complete", "", 0);
        });
        player.setOnErrorListener((mp, what, extra) -> {
            enqueueEvent(record, "error", "media_player_error_" + what + "_" + extra, what);
            return true;
        });

        if (isUriSource(source)) {
            player.setDataSource(context, Uri.parse(source));
        } else {
            player.setDataSource(source);
        }
        player.prepare();

        record.source = source;
        record.mediaPlayer = player;
    }

    private void enqueueEvent(PlayerRecord record, String type, String error, int code) {
        if (record == null) return;
        try {
            JSONObject event = buildPlayerJson(record.playerId);
            event.put("type", type == null ? "" : type);
            event.put("error", error == null ? "" : error);
            event.put("code", code);
            event.put("position", positionSeconds(record));
            String eventText = event.toString();
            if (record.pendingListener != null) {
                PendingListener pending = record.pendingListener;
                record.pendingListener = null;
                pending.result.success(pending.requestId, eventText);
            } else {
                record.events.add(eventText);
            }
        } catch (Throwable ignored) {
        }
    }

    private PlayerRecord requireRecord(LeafMethodCall call, Result result) throws Exception {
        return requireRecord(parseArgs(call.getArgs()), call, result);
    }

    private PlayerRecord requireRecord(JSONObject args, LeafMethodCall call, Result result) throws Exception {
        String playerId = requirePlayerId(args);
        PlayerRecord record = players.get(playerId);
        if (record == null) {
            result.error(call.getRequestId(), -404, "player_not_found", false);
            return null;
        }
        return record;
    }

    private String requirePlayerId(JSONObject args) throws Exception {
        String playerId = args.optString("playerId", "");
        if (playerId.isEmpty()) {
            throw new IllegalArgumentException("player_id_empty");
        }
        return playerId;
    }

    private JSONObject buildPlayerJson(String playerId) throws Exception {
        JSONObject out = new JSONObject();
        out.put("playerId", playerId == null ? "" : playerId);
        return out;
    }

    private JSONObject parseArgs(String args) throws Exception {
        if (args == null || args.trim().isEmpty()) {
            return new JSONObject();
        }
        return new JSONObject(args);
    }

    private String errorText(Throwable throwable) {
        if (throwable == null) {
            return "unknown_error";
        }
        String message = throwable.getMessage();
        if (message == null || message.isEmpty()) {
            return throwable.getClass().getSimpleName();
        }
        return message;
    }

    private boolean isUriSource(String source) {
        return source.startsWith("content://")
            || source.startsWith("android.resource://")
            || source.startsWith("file://")
            || source.startsWith("http://")
            || source.startsWith("https://");
    }

    private double durationSeconds(PlayerRecord record) {
        try {
            if (record == null || record.mediaPlayer == null) return 0.0;
            return Math.max(0, record.mediaPlayer.getDuration()) / 1000.0;
        } catch (Throwable ignored) {
            return 0.0;
        }
    }

    private double positionSeconds(PlayerRecord record) {
        try {
            if (record == null || record.mediaPlayer == null) return 0.0;
            return Math.max(0, record.mediaPlayer.getCurrentPosition()) / 1000.0;
        } catch (Throwable ignored) {
            return 0.0;
        }
    }

    private void releaseRecord(PlayerRecord record) {
        if (record == null) return;
        releaseMediaPlayerOnly(record);
        if (record.pendingListener != null) {
            record.pendingListener.result.error(record.pendingListener.requestId, -1, "player_disposed", true);
            record.pendingListener = null;
        }
        record.events.clear();
    }

    private void releaseMediaPlayerOnly(PlayerRecord record) {
        if (record == null || record.mediaPlayer == null) return;
        try {
            record.mediaPlayer.setOnCompletionListener(null);
            record.mediaPlayer.setOnErrorListener(null);
            record.mediaPlayer.release();
        } catch (Throwable ignored) {
        }
        record.mediaPlayer = null;
    }

    private static final class PendingListener {
        final int requestId;
        final Result result;

        PendingListener(int requestId, Result result) {
            this.requestId = requestId;
            this.result = result;
        }
    }

    private interface ThrowingRunnable {
        void run() throws Exception;
    }

    private static final class PlayerRecord {
        String playerId;
        String source;
        boolean looping;
        float volume = 1.0f;
        MediaPlayer mediaPlayer;
        PendingListener pendingListener;
        final Queue<String> events = new ArrayDeque<>();
    }
}

File: plugins/audio_player/ios/module/Classes/AudioPlayerPlugin.m
#import "AudioPlayerPlugin.h"
#import <AVFoundation/AVFoundation.h>

static const int32_t REQUEST_ERROR_INVALID_ARGS = -1;
static const int32_t REQUEST_ERROR_SOURCE_NOT_SET = -2;
static const int32_t REQUEST_ERROR_PLAYER_NOT_FOUND = -404;
static const int32_t REQUEST_ERROR_METHOD_NOT_IMPLEMENTED = -404;
static const int32_t REQUEST_ERROR_INTERNAL = -500;
static const int32_t REQUEST_ERROR_LISTEN_IN_PROGRESS = -11;

#pragma mark - AudioPlayerRecord

@interface AudioPlayerRecord : NSObject
@property (nonatomic, copy) NSString *playerId;
@property (nonatomic, copy) NSString *source;
@property (nonatomic, assign) BOOL looping;
@property (nonatomic, assign) float volume;
@property (nonatomic, strong, nullable) AVAudioPlayer *player;
@property (nonatomic, strong, nullable) id<LeafResult> pendingListener;
@property (nonatomic, assign) int32_t pendingListenerRequestId;
@property (nonatomic, strong) NSMutableArray<NSString *> *events;
@end

@implementation AudioPlayerRecord
- (instancetype)init {
    self = [super init];
    if (self) {
        _looping = NO;
        _volume = 1.0f;
        _events = [NSMutableArray array];
    }
    return self;
}
@end

#pragma mark - AudioPlayerPlugin

@interface AudioPlayerPlugin () <AVAudioPlayerDelegate>
@property (nonatomic, strong) NSMutableDictionary<NSString *, AudioPlayerRecord *> *players;
@end

@implementation AudioPlayerPlugin

- (instancetype)init {
    self = [super init];
    if (self) {
        _players = [NSMutableDictionary dictionary];
        [self configureAudioSession];
    }
    return self;
}

- (void)configureAudioSession {
    AVAudioSession *session = [AVAudioSession sharedInstance];
    NSError *error = nil;
    [session setCategory:AVAudioSessionCategoryPlayback error:&error];
    if (!error) {
        [session setActive:YES error:nil];
    }
}

#pragma mark - LeafPlugin

- (NSString *)pluginName {
    return @"AudioPlayerPlugin";
}

- (BOOL)canHandle:(NSString *)method {
    return [method hasPrefix:@"audio_player."];
}

- (void)onMethodCall:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    if (![NSThread isMainThread]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self onMethodCall:call result:result];
        });
        return;
    }

    NSString *method = call.method ?: @"";
    @try {
        if ([method isEqualToString:@"audio_player.set_source"]) {
            [self handleSetSource:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.play"]) {
            [self handlePlay:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.pause"]) {
            [self handlePause:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.stop"]) {
            [self handleStop:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.seek"]) {
            [self handleSeek:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.set_looping"]) {
            [self handleSetLooping:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.set_volume"]) {
            [self handleSetVolume:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.get_duration"]) {
            [self handleGetDuration:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.get_position"]) {
            [self handleGetPosition:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.is_playing"]) {
            [self handleIsPlaying:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.listen"]) {
            [self handleListen:call result:result];
            return;
        }
        if ([method isEqualToString:@"audio_player.dispose"]) {
            [self handleDispose:call result:result];
            return;
        }
        [result error:call.requestId code:REQUEST_ERROR_METHOD_NOT_IMPLEMENTED error:@"method_not_implemented" canceled:NO];
    } @catch (NSException *exception) {
        NSString *error = exception.reason.length > 0 ? exception.reason : @"plugin_dispatch_throw";
        [result error:call.requestId code:REQUEST_ERROR_INTERNAL error:error canceled:NO];
    }
}

#pragma mark - Method handlers

- (void)handleSetSource:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    NSString *playerId = [self requirePlayerId:args call:call result:result];
    if (!playerId) return;

    NSString *source = [args[@"source"] isKindOfClass:[NSString class]] ? args[@"source"] : @"";
    if (source.length == 0) {
        [result error:call.requestId code:REQUEST_ERROR_INVALID_ARGS error:@"source_empty" canceled:NO];
        return;
    }

    AudioPlayerRecord *record = [self getOrCreateRecord:playerId];
    [self releasePlayer:record];

    NSURL *url = [NSURL fileURLWithPath:source];
    NSError *error = nil;
    AVAudioPlayer *player = [[AVAudioPlayer alloc] initWithContentsOfURL:url error:&error];
    if (!player) {
        NSString *errorText = error.localizedDescription ?: @"audio_open_failed";
        [result error:call.requestId code:REQUEST_ERROR_INTERNAL error:errorText canceled:NO];
        return;
    }

    player.delegate = self;
    player.numberOfLoops = record.looping ? -1 : 0;
    player.volume = record.volume;
    [player prepareToPlay];

    record.source = source;
    record.player = player;
    [result success:call.requestId data:[self buildPlayerJson:playerId]];
}

- (void)handlePlay:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    if (!record.player) {
        [result error:call.requestId code:REQUEST_ERROR_SOURCE_NOT_SET error:@"source_not_set" canceled:NO];
        return;
    }
    [record.player play];
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handlePause:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    if (record.player && record.player.isPlaying) {
        [record.player pause];
    }
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handleStop:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    if (record.player) {
        [record.player stop];
        record.player.currentTime = 0.0;
    }
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handleSeek:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    AudioPlayerRecord *record = [self requireRecord:args call:call result:result];
    if (!record) return;
    double position = MAX(0.0, [args[@"position"] doubleValue]);
    if (record.player) {
        if (position > record.player.duration) {
            position = record.player.duration;
        }
        record.player.currentTime = position;
    }
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handleSetLooping:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    AudioPlayerRecord *record = [self requireRecord:args call:call result:result];
    if (!record) return;
    BOOL looping = [args[@"looping"] boolValue];
    record.looping = looping;
    if (record.player) {
        record.player.numberOfLoops = looping ? -1 : 0;
    }
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handleSetVolume:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    AudioPlayerRecord *record = [self requireRecord:args call:call result:result];
    if (!record) return;
    double volume = [args[@"volume"] doubleValue];
    if (volume < 0.0) volume = 0.0;
    if (volume > 1.0) volume = 1.0;
    record.volume = (float)volume;
    if (record.player) {
        record.player.volume = record.volume;
    }
    [result success:call.requestId data:[self buildPlayerJson:record.playerId]];
}

- (void)handleGetDuration:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    double duration = record.player ? record.player.duration : 0.0;
    NSString *json = [self jsonStringFromDictionary:@{
        @"playerId": record.playerId ?: @"",
        @"duration": @(duration)
    }];
    if (!json) {
        [result error:call.requestId code:REQUEST_ERROR_INTERNAL error:@"serialize_result_failed" canceled:NO];
        return;
    }
    [result success:call.requestId data:json];
}

- (void)handleGetPosition:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    double position = record.player ? record.player.currentTime : 0.0;
    NSString *json = [self jsonStringFromDictionary:@{
        @"playerId": record.playerId ?: @"",
        @"position": @(position)
    }];
    if (!json) {
        [result error:call.requestId code:REQUEST_ERROR_INTERNAL error:@"serialize_result_failed" canceled:NO];
        return;
    }
    [result success:call.requestId data:json];
}

- (void)handleIsPlaying:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    AudioPlayerRecord *record = [self requireRecord:call result:result];
    if (!record) return;
    BOOL playing = record.player ? record.player.isPlaying : NO;
    NSString *json = [self jsonStringFromDictionary:@{
        @"playerId": record.playerId ?: @"",
        @"playing": @(playing)
    }];
    if (!json) {
        [result error:call.requestId code:REQUEST_ERROR_INTERNAL error:@"serialize_result_failed" canceled:NO];
        return;
    }
    [result success:call.requestId data:json];
}

- (void)handleListen:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    NSString *playerId = [self requirePlayerId:args call:call result:result];
    if (!playerId) return;

    AudioPlayerRecord *record = [self getOrCreateRecord:playerId];

    if (record.pendingListener != nil) {
        [result error:call.requestId code:REQUEST_ERROR_LISTEN_IN_PROGRESS error:@"listen_in_progress" canceled:NO];
        return;
    }

    if (record.events.count > 0) {
        NSString *event = record.events.firstObject;
        [record.events removeObjectAtIndex:0];
        [result success:call.requestId data:event];
        return;
    }

    record.pendingListener = result;
    record.pendingListenerRequestId = call.requestId;
}

- (void)handleDispose:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSDictionary *args = [self parseArgs:call.args];
    NSString *playerId = [self requirePlayerId:args call:call result:result];
    if (!playerId) return;

    AudioPlayerRecord *record = self.players[playerId];
    if (record) {
        [self releaseRecord:record];
        [self.players removeObjectForKey:playerId];
    }
    [result success:call.requestId data:[self buildPlayerJson:playerId]];
}

#pragma mark - AVAudioPlayerDelegate

- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag {
    AudioPlayerRecord *record = [self recordForPlayer:player];
    if (!record) return;

    if (record.looping) {
        return;
    }

    if (!flag) {
        [self enqueueEvent:record type:@"error" error:@"audio_decode_error" code:-1];
        return;
    }

    double position = player.duration;
    [self enqueueEvent:record type:@"complete" error:@"" code:0 position:position];
}

- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError *)error {
    AudioPlayerRecord *record = [self recordForPlayer:player];
    if (!record) return;
    NSString *errorText = error.localizedDescription ?: @"decode_error";
    [self enqueueEvent:record type:@"error" error:errorText code:(int32_t)error.code];
}

#pragma mark - Record management

- (AudioPlayerRecord *)getOrCreateRecord:(NSString *)playerId {
    AudioPlayerRecord *existing = self.players[playerId];
    if (existing) {
        return existing;
    }
    AudioPlayerRecord *record = [[AudioPlayerRecord alloc] init];
    record.playerId = playerId;
    self.players[playerId] = record;
    return record;
}

- (AudioPlayerRecord *)requireRecord:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    return [self requireRecord:[self parseArgs:call.args] call:call result:result];
}

- (AudioPlayerRecord *)requireRecord:(NSDictionary *)args call:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSString *playerId = [self requirePlayerId:args call:call result:result];
    if (!playerId) return nil;
    AudioPlayerRecord *record = self.players[playerId];
    if (!record) {
        [result error:call.requestId code:REQUEST_ERROR_PLAYER_NOT_FOUND error:@"player_not_found" canceled:NO];
        return nil;
    }
    return record;
}

- (AudioPlayerRecord *)recordForPlayer:(AVAudioPlayer *)player {
    for (AudioPlayerRecord *record in self.players.allValues) {
        if (record.player == player) {
            return record;
        }
    }
    return nil;
}

- (AudioPlayerRecord *)recordForPlayerId:(NSString *)playerId {
    AudioPlayerRecord *record = self.players[playerId];
    if (!record) return nil;
    return record;
}

#pragma mark - Event queue

- (void)enqueueEvent:(AudioPlayerRecord *)record
                type:(NSString *)type
               error:(NSString *)error
                code:(int32_t)code {
    [self enqueueEvent:record type:type error:error code:code position:(record.player ? record.player.currentTime : 0.0)];
}

- (void)enqueueEvent:(AudioPlayerRecord *)record
                type:(NSString *)type
               error:(NSString *)error
                code:(int32_t)code
            position:(double)position {
    if (!record) return;

    NSString *json = [self jsonStringFromDictionary:@{
        @"playerId": record.playerId ?: @"",
        @"type": type ?: @"",
        @"error": error ?: @"",
        @"code": @(code),
        @"position": @(position)
    }];
    if (!json) return;

    if (record.pendingListener != nil) {
        id<LeafResult> pending = record.pendingListener;
        int32_t pendingId = record.pendingListenerRequestId;
        record.pendingListener = nil;
        record.pendingListenerRequestId = 0;
        [pending success:pendingId data:json];
    } else {
        [record.events addObject:json];
    }
}

#pragma mark - Lifecycle

- (void)releasePlayer:(AudioPlayerRecord *)record {
    if (!record || !record.player) return;
    record.player.delegate = nil;
    [record.player stop];
    record.player = nil;
}

- (void)releaseRecord:(AudioPlayerRecord *)record {
    if (!record) return;
    [self releasePlayer:record];
    if (record.pendingListener != nil) {
        [record.pendingListener error:record.pendingListenerRequestId
                                 code:REQUEST_ERROR_INTERNAL
                                error:@"player_disposed"
                             canceled:YES];
        record.pendingListener = nil;
        record.pendingListenerRequestId = 0;
    }
    [record.events removeAllObjects];
}

#pragma mark - Helpers

- (NSString *)requirePlayerId:(NSDictionary *)args call:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSString *playerId = [args[@"playerId"] isKindOfClass:[NSString class]] ? args[@"playerId"] : @"";
    if (playerId.length == 0) {
        [result error:call.requestId code:REQUEST_ERROR_INVALID_ARGS error:@"player_id_empty" canceled:NO];
        return nil;
    }
    return playerId;
}

- (NSString *)buildPlayerJson:(NSString *)playerId {
    return [self jsonStringFromDictionary:@{
        @"playerId": playerId ?: @""
    }] ?: @"{}";
}

- (NSDictionary *)parseArgs:(NSString *)args {
    if (args.length == 0) {
        return @{};
    }
    NSData *data = [args dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) {
        return nil;
    }
    id json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    if (![json isKindOfClass:[NSDictionary class]]) {
        return nil;
    }
    return (NSDictionary *)json;
}

- (NSString *)jsonStringFromDictionary:(NSDictionary *)dictionary {
    NSError *error = nil;
    NSData *data = [NSJSONSerialization dataWithJSONObject:dictionary options:0 error:&error];
    if (!data || error) {
        return nil;
    }
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end
